app/src/main/java/com/quran/labs/androidquran/widgets/WidgetListProvider.kt

```diff
@@ -0,0 +1,77 @@
+package com.quran.labs.androidquran.widgets
+
+import android.content.Context
+import android.content.Intent
+import android.os.Bundle
+import android.widget.RemoteViews
+import android.widget.RemoteViewsService.RemoteViewsFactory
+import com.quran.data.core.QuranInfo
+import com.quran.labs.androidquran.QuranApplication
+import com.quran.labs.androidquran.R
+import com.quran.labs.androidquran.data.QuranDisplayData
+import com.quran.labs.androidquran.database.BookmarksDBAdapter
+import javax.inject.Inject
+
+class WidgetListProvider(private val context: Context) : RemoteViewsFactory {
+
+  private var suraPageItemList: List<SuraPageItem> = listOf()
+
+  @Inject
+  lateinit var quranInfo: QuranInfo
+
+  @Inject
+  lateinit var quranDisplayData: QuranDisplayData
+
+  init {
+    (context.applicationContext as QuranApplication).applicationComponent.inject(this)
+    populateListItem()
+  }
+
+  private fun populateListItem() {
+    val appContext = context.applicationContext
+    val mBookmarksDBAdapter = BookmarksDBAdapter(appContext, quranInfo.numberOfPages)
+    val bookmarksList = mBookmarksDBAdapter.getBookmarks(BookmarksDBAdapter.SORT_DATE_ADDED)
+    suraPageItemList = bookmarksList.filter { it.isPageBookmark() }.map {
+      SuraPageItem(sura = quranDisplayData.getSuraNameString(appContext, it.page), page = it.page)
+    }
+  }
+
+  override fun onCreate() = Unit
+  override fun onDestroy() = Unit
+
+  override fun onDataSetChanged() {
+    populateListItem()
+  }
+
+  override fun getCount() = suraPageItemList.size
+
+  override fun getItemId(position: Int) = position.toLong()
+
+  override fun hasStableIds() = false
+
+  /*
+  *Similar to getView of Adapter where instead of View
+  *we return RemoteViews
+  *
+  */
+  override fun getViewAt(position: Int): RemoteViews {
```

Comment:
is there no recycling that we can get with this?

Reply:
I don't think so, [looking at the interface methods(https://developer.android.com/reference/android/widget/RemoteViewsService.RemoteViewsFactory#getViewAt(int)). It doesn't seem like the [`Adapter#getView`](https://developer.android.com/reference/android/widget/Adapter#getView(int,%20android.view.View,%20android.view.ViewGroup)) method.